File: Day4.playground/Sources/Day4.swift
import Foundation

let inputs = (146810 ... 612564).map { $0.digits }

extension Int {
    var digits: [Int] {
        String(self).compactMap { $0.wholeNumberValue }
    }
}

extension Array where Element == Int {
    var isIncreasing: Bool {
        var cursor = 0
        var result = true

        for digit in self {
            if digit < cursor {
                result = false
                break
            }
            cursor = digit
        }

        return result
    }

    var hasDouble: Bool {
        var cursor = self[0]
        var result = false

        for digit in self.dropFirst() {
            if digit == cursor {
                result = true
                break
            }
            cursor = digit
        }

        return result
    }

    var hasStrictDouble: Bool {
        func aggregate(_ dict: inout [Int: Int], _ digit: Int) {
            dict[digit, default: 0] += 1
        }

        return self.reduce(into: [:], aggregate)
            .values
            .contains(2)
    }
}

func isValid(_ digits: [Int]) -> Bool {
    digits.isIncreasing && digits.hasDouble
}

func isStrictlyValid(_ digits: [Int]) -> Bool {
    digits.isIncreasing && digits.hasStrictDouble
}


public let part1 = inputs
    .filter(isValid)
    .count

public let part2 = inputs
    .filter(isStrictlyValid)
    .count

File: Day4.hs
-- | Advent of Code 2019 Day p

module Main where

import Data.List (groupBy)
  
type Digits = [Int]
  
toDigits :: Int -> Digits
toDigits 0 = []
toDigits x = toDigits (x `div` 10) ++ [x `mod` 10]


isIncreasing :: Digits -> Bool
isIncreasing (x:y:[]) = y >= x
isIncreasing (x:y:ys) = if y >= x
                        then isIncreasing (y:ys)
                        else False

                        
hasDouble :: Digits -> Bool
hasDouble (x:y:[]) = x == y
hasDouble (x:y:ys) = if x == y
                     then True
                     else hasDouble (y:ys)
                     
isValid :: Digits -> Bool
isValid d = (hasDouble d) && (isIncreasing d)
 
inputs :: [Digits] 
inputs = map toDigits [146810 .. 612564]

validPasswords = filter isValid inputs

hasDouble' :: Digits -> Bool
hasDouble' d = 2 `elem` grouped
  where grouped = map length $ groupBy (==) d


isValid' :: Digits -> Bool
isValid' d = (hasDouble' d) && (isIncreasing d)
  
validPasswords' = filter isValid' validPasswords

part1 = show $ length $ validPasswords
part2 = show $ length $ validPasswords'

main :: IO ()
main =
  putStrLn $ "Part1: " ++ part1 ++ "\n" ++ "Part2: " ++ part2

